WMF Framework/EventLoggingService.swift

```diff
@@ -286,38 +288,32 @@ public class EventLoggingService : NSObject, URLSessionDelegate {
     
     @objc
     private func tryPostEvents() {
+        let operation = AsyncBlockOperation { (operation) in
+            let moc = self.managedObjectContext
+            moc.perform {
+                let fetch: NSFetchRequest<EventRecord> = EventRecord.fetchRequest()
+                fetch.sortDescriptors = [NSSortDescriptor(keyPath: \EventRecord.recorded, ascending: true)]
+                fetch.predicate = NSPredicate(format: "(posted == nil) AND (failed != TRUE)")
+                fetch.fetchLimit = self.postBatchSize
 
-        self.postLock.lock()
-        guard started, !posting else {
-            self.postLock.unlock()
-            return
-        }
-        posting = true
-        self.postLock.unlock()
-        
-        let moc = self.managedObjectContext
-        moc.perform {
-            let fetch: NSFetchRequest<EventRecord> = EventRecord.fetchRequest()
-            fetch.sortDescriptors = [NSSortDescriptor(keyPath: \EventRecord.recorded, ascending: true)]
-            fetch.predicate = NSPredicate(format: "(posted == nil) AND (failed != TRUE)")
-            fetch.fetchLimit = self.postBatchSize
-            
-            do {
-                var eventRecords: [EventRecord] = []
-                defer {
-                    if eventRecords.count > 0 {
-                        self.postEvents(eventRecords)
-                    } else {
-                        self.postLock.lock()
-                        self.posting = false
-                        self.postLock.unlock()
+                do {
+                    var eventRecords: [EventRecord] = []
+                    defer {
+                        if eventRecords.count > 0 {
+                            self.postEvents(eventRecords, completion: {
+                                operation.finish()
+                            })
+                        }
                     }
+                    eventRecords = try moc.fetch(fetch)
+                } catch let error {
+                    DDLogError(error.localizedDescription)
                 }
-                eventRecords = try moc.fetch(fetch)
-            } catch let error {
-                DDLogError(error.localizedDescription)
+
+                operation.finish()
```

Comment:
This will finish early if there are event records. This might be cleaner with the event record declaration outside of the try catch, and the if statement at the end of the method instead of inside the defer as 
```
if eventRecords.count > 0 {
   self.postEvents(eventRecords, completion: {
    operation.finish()
  })
} else {
  operation.finish()
}